### PR TITLE
Clamp floats

### LINKED ISSUE #848: Serializing sh-float is hard
The current definition of sh-float is surprisingly hard to serialize correctly, the best C-code I could come up with is:

        int
        serialize_sh_float(char buf[18], double d)
        {
                int p = 0, l, dig;

                buf[0] = '*';
                buf[1] = '\0';
                if (isnan(d))
                        return (-1);
                if (fabs(d) >= 99999999999999.95)
                        return (-1);

                if (d < 0) {
                        buf[p++] = '-';
                        d = -d;
                        l = 18;
                } else {
                        l = 17;
                }
                dig = trunc(log10(d));
                p += snprintf(buf + p, l - p, "%.*f", 15-dig, d);
                while (buf[p-2] == '0') {
                        buf[p-1] = '\0';
                        p--;
                }
                return (p);
        }

And that is horrible in severale levels.

If we insist instead that sh-float can have maximum six digits after the decimal point, people can rely on the default "6 decimals" behaviour of "%f" formatters, and use the simpler:

        int
        serialize_sh_float(char buf[18], double d)
        {
                int p = 0, l, dig;

                if (isnan(d))
                        return (-1);
                if (fabs(d) >= 99999999999999.95)
                        return (-1);

                if (d < 0) {
                        buf[p++] = '-';
                        d = -d;
                }
                p += snprintf(buf + p, 18 - p, "%-16f", d);
                while (buf[p-1] == ' ')
                        p -= 1;
                buf[p] = '\0';
                return (p);
        }

### PR DESCRIPTION
**Pull request by mnot**
Fixes #848

### CODE REVIEW

**Comment by phluid61** [comment]
Is this PR meant to include #892?

**Comment by mnot** [comment]
Yeah, oops. Need more coffee.

**Comment by mnot** [comment]
Try now.

File: draft-ietf-httpbis-header-structure.md [hunk 1]
```diff
@@ -856,6 +851,8 @@ _RFC Editor: Please remove this section before publication._
 ## Since draft-ietf-httpbis-header-structure-11
 
 * Allow \* in key (#844).
+* Allow dictionary members to have parameters (#842).
```
Comment (phluid61) [inline draft-ietf-httpbis-header-structure.md]: Close enough

**Comment by phluid61** [comment]
Otherwise, looks alright to me.

File: draft-ietf-httpbis-header-structure.md [hunk 2]
```diff
@@ -498,13 +491,14 @@ Given an integer as input_integer:
...
 
-0. If input_float is not a IEEE 754 double precision number, fail serialisation.
-1. Let output be an empty string.
-2. If input_float is less than (but not equal to) 0, append "-" to output.
-3. Append input_float's integer component represented in base 10 using only decimal digits to output; if it is zero, append "0".
-4. Append "." to output.
-5. Append input_float's decimal component represented in base 10 using only decimal digits to output; if it is zero, append "0".
-6. Return output.
+0. If input_float's fractional part has more than six digits of precision, fail serialisation.
+1. If the number of digits of precision in input_float's fractional part plus those in its integer part add to more than fifteen digits, fail serialisation.
```
Comment (bsdphk) [inline draft-ietf-httpbis-header-structure.md]: Not sure I agree with this one, what if the value is 1/3 ?

Comment (mnot) [reply draft-ietf-httpbis-header-structure.md]: Perhaps something like:

> Append up to six digits of input_float's decimal component...

Comment (phluid61) [reply draft-ietf-httpbis-header-structure.md]: 1/3 doesn't match "integers with a fractional part, which can be at most six digits of precision." This isn't the algorithm for serialising an IEEE-754 float, nor an arbitrary rational number.

Although maybe guidance for converting those to an SH-float data object could be useful.

File: draft-ietf-httpbis-header-structure.md [hunk 3]
```diff
@@ -278,26 +278,19 @@ Note that commas in integers are used in this section's prose only for readabili
 
 ## Floats {#float}
 
-Floats are integers with a fractional part, that can be stored as IEEE 754 double precision numbers (binary64) ({{IEEE754}}).
+Floats are integers with a fractional part, which can be at most six digits of precision. Additionally, like integers, it can have no more than fifteen digits in total, which in some cases further constrains its precision.
```
Comment (phluid61) [inline draft-ietf-httpbis-header-structure.md]: Might be an idea to say "decimal digits" here.

Comment (ioggstream) [reply draft-ietf-httpbis-header-structure.md]: > Floats are integers with a fractional part

I would prefere "floats are numbers with a fractional part"

Comment (phluid61) [reply draft-ietf-httpbis-header-structure.md]: Or "numbers with integer and fractional parts"

I still think saying "decimal digits" would help, too.

File: draft-ietf-httpbis-header-structure.md [hunk 4]
```diff
@@ -278,7 +278,7 @@ Note that commas in integers are used in this section's prose only for readabili
 
 ## Floats {#float}
 
-Floats are integers with a fractional part, which can be at most six digits of precision. Additionally, like integers, it can have no more than fifteen digits in total, which in some cases further constrains its precision.
+Floats are decimal numbers with an integer and a fractional part. The fractional part has at most six digits of precision. Additionally, like integers, it can have no more than fifteen digits in total, which in some cases further constrains its precision.
```
Comment (ioggstream) [inline draft-ietf-httpbis-header-structure.md]: LGTM

File: draft-ietf-httpbis-header-structure.md [hunk 5]
```diff
@@ -520,13 +513,14 @@ Given an integer as input_integer:
...
 Given a float as input_float:
 
-0. If input_float is not a IEEE 754 double precision number, fail serialisation.
-1. Let output be an empty string.
-2. If input_float is less than (but not equal to) 0, append "-" to output.
-3. Append input_float's integer component represented in base 10 using only decimal digits to output; if it is zero, append "0".
-4. Append "." to output.
-5. Append input_float's decimal component represented in base 10 using only decimal digits to output; if it is zero, append "0".
-6. Return output.
+0. If input_float's fractional component has more than six digits of precision, fail serialisation.
```
Comment (ioggstream) [inline draft-ietf-httpbis-header-structure.md]: Q: here we use fractional component, above we use fractional part. Is that ok?

Comment (mnot) [reply draft-ietf-httpbis-header-structure.md]: Good catch.

**Pull request merged**